src/components/projects/grants/GrantsForm.tsx

```diff
@@ -0,0 +1,286 @@
+"use client"
+
+import { useFieldArray, useForm } from "react-hook-form"
+import { z } from "zod"
+import { zodResolver } from "@hookform/resolvers/zod"
+import { Plus } from "lucide-react"
+import { useState } from "react"
+import { Checkbox } from "@/components/ui/checkbox"
+import { cn } from "@/lib/utils"
+import { Form } from "@/components/ui/form"
+import { Button } from "@/components/ui/button"
+import { FUNDING_TYPES, FundingFormSchema, FundingType } from "./schema"
+import {
+  OptimismFundingForm,
+  OtherFundingForm,
+  VentureFundingForm,
+} from "./FundingForm"
+
+export const GrantsForm = () => {
+  const form = useForm<z.infer<typeof FundingFormSchema>>({
+    resolver: zodResolver(FundingFormSchema),
+    mode: "onSubmit",
+    reValidateMode: "onChange",
+    defaultValues: {
+      venture: [],
+      optimism: [],
+      other: [],
+    },
+  })
+
+  const [selectedNone, setSelectedNone] = useState(false)
+
+  const {
+    fields: ventureFields,
+    replace: setVentureFields,
+    append: addVentureField,
+  } = useFieldArray({
+    control: form.control,
+    name: "venture",
+  })
+
+  const {
+    fields: optimismFields,
+    replace: setOptimismFields,
+    append: addOptimismField,
+  } = useFieldArray({
+    control: form.control,
+    name: "optimism",
+  })
+
+  const {
+    fields: otherFields,
+    replace: setOtherFields,
+    append: addOtherField,
+  } = useFieldArray({
+    control: form.control,
+    name: "other",
+  })
+
+  const hasFundingType = (type: FundingType) => {
+    switch (type) {
+      case "venture":
+        return ventureFields.length > 0
+      case "optimism":
+        return optimismFields.length > 0
+      case "other":
+        return otherFields.length > 0
+      case "none":
+        return selectedNone
+      default:
+        return false
+    }
+  }
+
+  const onToggleFundingType = (type: FundingType) => {
+    if (type === "venture") {
+      if (ventureFields.length === 0) {
+        setSelectedNone(false)
+        addVentureField({
+          amount: "",
+          year: "",
+          details: "",
+        })
+      } else {
+        setVentureFields([])
+      }
+    } else if (type === "optimism") {
+      if (optimismFields.length === 0) {
+        setSelectedNone(false)
+        addOptimismField({
+          type: "",
+          link: "",
+          amount: "" as unknown as number, // We coerce it with zod, so this is a better empty UI
```

Comment:
can it do `undefined` as empty state instead?

Reply:
no bc it's a required field 😵‍💫